### PR TITLE
Sto email notifications doc 2800

### PR DESCRIPTION
**Pull request by douglas-j-bothwell**
# Harness Developer Pull Request
Thanks for helping us make the Developer Hub better. The PR will be looked at
by the maintainers. 

## What Type of PR is This?

- [ ] Issue
- [ ] Feature
- [ ] Maintenance/Chore

If tied to an Issue, list the Issue(s) here:

* *Issue(s)*

## House Keeping
Some items to keep track of. Screen shots of changes are optional but would help the maintainers review quicker. 

- [ ] Tested Locally
- [ ] *Optional* Screen Shoot.

### CODE REVIEW

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 1]
```diff
@@ -0,0 +1,279 @@
+---
+title: Notification workflows -- Automated emails for detected issues
+description: Send emails and stop pipelines automatically based on severity levels
+sidebar_position: 210
+---
+
+This topic describes two example workflows for setting up automatic notifications based on scan results. 
+
+When you run a scan, the step generates a set of of output variables that capture the number of issues detected at each severity level: CRITICAL, HIGH, MEDIUM, and so on. You can [include these variables in automated notifications](#send-notification-emails-with-scan-results) such as emails. You can also [create governance policies that evaluate these variables](#stop-builds-and-generate-notifications-using-governance-policies) and use the results to stop pipeline builds and send automated notifications.
```
Comment (r0610205) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:9]: `a set of of output variables` - of repeated twice

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+
+## Send notifications with scan results
+
+In this workflow, you add an Email step that notifies recipients of the the scan results. The email lists the number of issues detected, grouped by severity level: critical, high, medium, and so on. The pipeline sends a notification whenever the previous scan step finishes successfully.
```
Comment (r0610205) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:14]: `recipients of the the scan results` - the is repeated twice

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+In this workflow, you create a simple OPA policy for the pipeline: If the scan detected any NEW_CRITICAL or NEW_HIGH severities, exit the build with an error and send an email. "New-severity" issues are issues that were found in the scanned target but not the baseline —  or, if the target has no baseline, the previous scan. 
+
+
+### Create the OPA policy
+
+1. Go to **Account Settings** (left menu) &gt; **Policies**.
+
+2. Click **Policies** (top right) and then **New Policiy**. 
+
+3. Given your new policy the name **Security no NEW_CRITICAL or NEW_HIGH issues**.
```
Comment (r0610205) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:58]: Should be `Give`

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+                    body: |-
+                      "STO scan of <+pipeline.name> found the following issues: <br> 
+                       Critical : <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.CRITICAL> <br>
+                       New Critical : <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.NEW_CRITICAL> <br>
+                       High: <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.HIGH> &#10; <br>
+                       New High: <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.NEW_HIGH> <br>
+                       Medium: <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.MEDIUM> <br>
+                       New Medium: <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.NEW_MEDIUM>  <br>
+                       See https://app.harness.io/ng/#/account/H5W8iol5TNWc4G9h5A2MXg/sto/orgs/default/"
+                  timeout: 1d
```
Comment (r0610205) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:246]: I'd replace hardcoded account Id with something like <Your Account ID> to make it clear they need to modify it

**Review by r0610205 — APPROVED** [review]

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 5]
```diff
@@ -0,0 +1,279 @@
+---
+title: Notification workflows -- Automated emails for detected issues
+description: Send emails and stop pipelines automatically based on severity levels
+sidebar_position: 210
+---
+
+This topic describes two example workflows for setting up automatic notifications based on scan results. 
+
+When you run a scan, the step generates a set of output variables that capture the number of issues detected at each severity level: CRITICAL, HIGH, MEDIUM, and so on. You can [include these variables in automated notifications](#send-notification-emails-with-scan-results) such as emails. You can also [create governance policies that evaluate these variables](#stop-builds-and-generate-notifications-using-governance-policies) and use the results to stop pipeline builds and send automated notifications.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:9]:
```suggestion
When you run a scan, the step generates a set of output variables that capture the number of issues detected at each severity level: CRITICAL, HIGH, MEDIUM, and so on. You can [include these variables in automated notifications](#send-notification-emails-with-scan-results), such as emails. You can also [create governance policies that evaluate these variables](#stop-builds-and-generate-notifications-using-governance-policies) and use the results to stop pipeline builds and send automated notifications.
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+
+
+## Send notifications with scan results
+
+In this workflow, you add an Email step that notifies recipients of the scan results. The email lists the number of issues detected, grouped by severity level: critical, high, medium, and so on. The pipeline sends a notification whenever the previous scan step finishes successfully.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:14]: In the 2nd paragraph, the severity levels were capitalized but here, they are lowercase.

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+             Critical : <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.CRITICAL> <br>
+             New Critical : <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.NEW_CRITICAL> <br>
+             High: <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.HIGH> <br>
+             New High: <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.NEW_HIGH> <br>
+             Medium: <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.MEDIUM> <br>
+             New Medium: <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.NEW_MEDIUM> <br> 
+             See https://app.harness.io/ng/#/account/ACCOUNT_ID/sto/orgs/default/"
+      timeout: 1d
+```
+For more information about pipeline and other variables, go to [Built-in and Custom Harness Variables Reference](/docs/platform/Variables-and-Expressions/harness-variables).
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:40]:
```suggestion
For more information about pipeline variables and other variables, go to [Built-in and Custom Harness Variables Reference](/docs/platform/Variables-and-Expressions/harness-variables).
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+
+
+## Stop builds and generate notifications using governance policies
+
+:::note
+This workflow requires a basic knowledge of governance policies and how to implement them using [Harness Policy as Code](/docs/platform/Policy-as-code/harness-governance-overview) and [Open Policy Agent (OPA)](https://www.openpolicyagent.org/)
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:46]:
```suggestion
This workflow requires a basic knowledge of governance policies and how to implement them using [Harness Policy as Code](/docs/platform/Policy-as-code/harness-governance-overview) and [Open Policy Agent (OPA)](https://www.openpolicyagent.org/).
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+:::
+
+In this workflow, you create a simple OPA policy for the pipeline: If the scan detected any NEW_CRITICAL or NEW_HIGH severities, exit the build with an error and send an email. "New-severity" issues are issues that were found in the scanned target but not the baseline —  or, if the target has no baseline, the previous scan. 
+
+
+### Create the OPA policy
+
+1. Go to **Account Settings** (left menu) &gt; **Policies**.
+
+2. Click **Policies** (top right) and then **New Policiy**.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:56]:
```suggestion
2. Select **Policies** (top right) and then **New Policy**. 
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+In this workflow, you create a simple OPA policy for the pipeline: If the scan detected any NEW_CRITICAL or NEW_HIGH severities, exit the build with an error and send an email. "New-severity" issues are issues that were found in the scanned target but not the baseline —  or, if the target has no baseline, the previous scan. 
+
+
+### Create the OPA policy
+
+1. Go to **Account Settings** (left menu) &gt; **Policies**.
+
+2. Click **Policies** (top right) and then **New Policiy**. 
+
+3. Give your new policy the name **Security no NEW_CRITICAL or NEW_HIGH issues**.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:58]:
```suggestion
3. Name the policy **Security no NEW_CRITICAL or NEW_HIGH issues**. 
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+For more information about pipeline and other variables, go to [Built-in and Custom Harness Variables Reference](/docs/platform/Variables-and-Expressions/harness-variables).
+
+
+## Stop builds and generate notifications using governance policies
+
+:::note
+This workflow requires a basic knowledge of governance policies and how to implement them using [Harness Policy as Code](/docs/platform/Policy-as-code/harness-governance-overview) and [Open Policy Agent (OPA)](https://www.openpolicyagent.org/)
+:::
+
+In this workflow, you create a simple OPA policy for the pipeline: If the scan detected any NEW_CRITICAL or NEW_HIGH severities, exit the build with an error and send an email. "New-severity" issues are issues that were found in the scanned target but not the baseline —  or, if the target has no baseline, the previous scan.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:49]:
```suggestion
In this workflow, you create a simple OPA policy for the pipeline: If the scan detected any NEW_CRITICAL or NEW_HIGH severities, exit the build with an error and send an email. _New-severity_ issues are issues that were found in the scanned target but not the baseline —  or, if the target has no baseline, the previous scan. 
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+   deny[sprintf("Scan results can't contain any NEW_HIGH vulnerabilities '%s'", [input])] {
+       input.NEW_HIGH != 0 
+   }
+   ```
+
+5. Click **Save**. 
+
+### Create a policy set
+
+A _policy set_ is simply a collection of one or more policies. You combine policies into a set and then include it in a Policy step.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:78]:
```suggestion
A _policy set_ is a collection of one or more policies. You combine policies into a set and then include it in a Policy step. 
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+   }
+   ```
+
+5. Click **Save**. 
+
+### Create a policy set
+
+A _policy set_ is simply a collection of one or more policies. You combine policies into a set and then include it in a Policy step. 
+
+1. In the Policies Overview page, click **Policy Sets** (top right) and then **New Policiy Set**.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:80]:
```suggestion
1. In the Policies Overview page, select **Policy Sets** (top right) and then **New Policy Set**. 
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+
+5. Click **Save**. 
+
+### Create a policy set
+
+A _policy set_ is simply a collection of one or more policies. You combine policies into a set and then include it in a Policy step. 
+
+1. In the Policies Overview page, click **Policy Sets** (top right) and then **New Policiy Set**. 
+
+2. Click **New Policy Set**. The Policy Set wizard appears.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:82]:
```suggestion
2. Select **New Policy Set**. The Policy Set wizard appears.
```

Is this a duplicate step or do you select **New Policy set** twice?

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+
+    2. Entity type this policy applies to = **Custom**
+
+    3. On what event should the policy be set to = **On Step**. 
+   
+       These settings allow you to apply the member policies to a specific step, which you'll define below. 
+
+4. Policy evaluation criteria: 
+ 
+   1. Click **Add Policy**.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:96]:
```suggestion
   1. Select **Add Policy**.
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+    
+   2. Select the policy you just created and set the pull-down to **Error and Exit**. This is the action to take if any policies in the set are violated. 
+
+     ![](./static/notif-opa-select-policy.png)
+
+   3. Click **Apply** to add the policy to the set, then **Finish** to close the Policy Set wizard.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:102]:
```suggestion
   3. Select **Apply** to add the policy to the set, then **Finish** to close the Policy Set wizard.
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+
+   4. :exclamation: In the **Policy Sets** page, enable  **Enforced** for your new policy set.
+
+      ![](./static/notif-opa-policy-set-enforced-yes.png)
+
+
+### Add a step to evaluate your policy against the scan results
+
+Now you will create a Policy step that applies the policy you just defined against the output variables from the scan. If the scan detected any NEW_CRITICAL or NEW_HIGH issues, the pipeline fails.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:111]:
```suggestion
Now you will create a Policy step that applies the policy set you just defined against the output variables from the scan. If the scan detected any NEW_CRITICAL or NEW_HIGH issues, the pipeline fails. 
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+
+2. Add a **Policy** step to the stage.
+
+3. Configure the step as follows:
+
+   1. Entity Type = **Custom**
+
+   2. Policy Set — Add the policy set you just created. 
+
+   3. Payload — Enter the following payload. Make sure that you replace `SCAN_STAGE_ID` and `SCAN_STEP_ID` with the stage and step IDs in your pipeline
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:123]:
```suggestion
   1. Entity Type = **Custom**

   2. Policy Set = Add the policy set you just created. 

   3. Payload = Enter the following payload. Make sure that you replace `SCAN_STAGE_ID` and `SCAN_STEP_ID` with the stage and step IDs in your pipeline
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+      }
+      ```
+
+
+### Set up email notifications for pipeline failures
+
+You have a Policy that fails the pipeline based on an OPA policy. Now you can configure the stage to send an email notification automatically whenever the pipeline fails. 
+
+
+1. Click **Notifications** (right-side menu). The New Notification wizard appears.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:138]:
```suggestion
1. Select **Notifications** (right-side menu). The New Notification wizard appears.
```

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+
+2. Set up the notification as follows:
+
+     1. Overview page —  Enter a notifcation name such as **Pipeline failed -- NEW_CRITICAL or NEW_HIGH issues detected**.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:142]:
```suggestion
     1. Overview page —  Enter a notification name, such as **Pipeline failed - NEW_CRITICAL or NEW_HIGH issues detected**.
```

There is a mix of = and - in lists in this topic.

File: docs/security-testing-orchestration/use-sto/email-notifications.md [hunk 21]
```diff
@@ -0,0 +1,279 @@
...
+
+     2. Pipeline Events page  —  Select **Stage Failed**, for the event that triggers the notification, and the stage that has the Policy step you just created.
```
Comment (aimurphy) [inline docs/security-testing-orchestration/use-sto/email-notifications.md:144]: Confused by the comma usage. Sentence is not clear.

**Comment by douglas-j-bothwell** [comment]
Closing this PR. Implemented AM's edits in https://github.com/harness/developer-hub/pull/1138

**Pull request closed without merging**